### PR TITLE
feature(automatic-transactions): Creates a decorator to make transactions easier for the developer

### LINKED ISSUE #75: Automatic transactions
## 🚀 Feature Proposal

In most applications I've seen:

```js
fastify.get('/something', function (req, reply) {
  return this.pg.transact(async (client) => {
      const { rows } = await client.query('SELECT id, username, hash, salt FROM users WHERE id=$1', [req.params.id])
      return rows
  })
})
```

I think we could greatly simplify this by having some dedicated syntax to avoid developer errors

### 3rd argument

```js
fastify.get('/something', { pg: { transact: true } }, function (req, reply, client) {
  const { rows } = await client.query('SELECT id, username, hash, salt FROM users WHERE id=$1', [req.params.id])
  return rows
})
```

or even better

```js
fastify.register(require('fastify-postgres'), {
  connectionString: 'postgres://postgres@localhost/postgres',
  wrapRoutesWithTransaction: true
})

fastify.get('/something', function (req, reply, client) {
  const { rows } = await client.query('SELECT id, username, hash, salt FROM users WHERE id=$1', [req.params.id])
  return rows
})
```

This can be implemented by having an `'onRoute'` hook that wraps the handler.

### Decorator

```js
fastify.get('/something', { pg: { transact: true } }, function (req, reply) {
  const { rows } = await req.pg.query('SELECT id, username, hash, salt FROM users WHERE id=$1', [req.params.id])
  return rows
})
```

or even better

```js
fastify.register(require('fastify-postgres'), {
  connectionString: 'postgres://postgres@localhost/postgres',
  wrapRoutesWithTransaction: true
})

fastify.get('/something', function (req, reply) {
  const { rows } = await req.pg.query('SELECT id, username, hash, salt FROM users WHERE id=$1', [req.params.id])
  return rows
})
```

This can be implemented by having a `'preHandler'` hook that starts the transaction and an `'onSend'` hook that commits or roll back it.

---

What do you prefer? The decorator is probably the more idiomatic approach, even it it will require a bit more work to implement.

### PR DESCRIPTION
**Pull request by toni-sharpe**
#### Description

* Enables handlers to be wrapped in a transaction which reduces developer boilerplate when defining routes.
* To do this we initialise some request hooks in the `onRoute` handler.
** `preHandler`: gets a client from the pool and decorates the request object with that client. Note that we use a client, not the db object that is at `fastify.pg`.
** `onError`: if the handler fails we rollback in this handler (and set a Boolean so we don't try to commit in onSend.
** `onSend`: We commit here if the handler is successful and make sure the client is released to the pool.
* We also handle namespacing, ensuring that the name provided in the plugin config is also used to decorate request with a client at that name.
* We use `{ pg: { transact: true || 'string' } }` to decide whether to use `req.pg` or `req.pg[name]` in the request hooks. Note that this was easier to handle than the proposed API in the issue.
* Various checks avoid adding handlers where they're not required.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

#### Questions

- The issue with what to do when a `try/catch` is used in the handler is still outstanding

Resolves #75

### CODE REVIEW

File: test/req-transaction.test.js
Comment (nherment) [inline test/req-transaction.test.js:27]: add option `{ pg: { transact: true } }`

**Comment by nherment** [comment]
What I think is missing is:
- 
- Ensure the new Fastify 'start' hook is triggered based on `{ pg: { transact: true } }`
- Ensure the hook that replaces req.pg with a single client, within a transaction (do a `lient.query('BEGIN')` as the `transact()` func does)
- Ensure the TX is closed (commited or rolled back) by the hook when the route handler is done

You can priobably use the existign `transact` method defined in the plugin

**Review by simoneb — COMMENTED** [review]
I'm posting this late but we went through this already anyway

File: index.js
Comment (simoneb) [inline index.js:66]: this can be written in a nicer way without reassigning variables and returning early

File: index.js
Comment (simoneb) [inline index.js:119]: are we allowing two different ways to require a transaction? why?

File: index.js
Comment (simoneb) [inline index.js:124]: prefer to use a symbol so this plugin is the only one who has access to it

File: index.js
Comment (simoneb) [inline index.js:130]: what does this do?

File: index.js
Comment (simoneb) [inline index.js:138]: why are we grabbing a client from the pool before knowing if we need a transaction? or actually, don't we already know, if we're executing this code, that we need a transaction so the next conditional is unnecessary?

File: index.js
Comment (simoneb) [inline index.js:153]: do we need this conditional? aren't we executing this code only if we already knew that we needed a transaction?

**Review by simoneb — CHANGES_REQUESTED** [review]

File: index.js
Comment (toni-sharpe) [reply index.js:119]: I think this is a misunderstanding on my part about how Fastify works.

When I was using the following syntax I was using an options key

```
method: 'GET',
options: { useTransaction },
handler ...
```

But I think I could put the key directly into that object in which would negate the need for the extra conditions

I will change

File: index.js
Comment (toni-sharpe) [reply index.js:130]: This is the best I could come up with in the response to the question 'what happens if rollback fails'

File: index.js
Comment (toni-sharpe) [reply index.js:138]: OK, so we can run all of this conditionally?

I was thinking we would still need to get the client in any case, but now I see that was a mistaken thought. This also relates to https://github.com/fastify/fastify-postgres/pull/76#discussion_r577157973 where I wanted to ensutre that I always released the client

It looks like the whole lot can go in to the conditional so I'll make that change

File: index.js
Comment (simoneb) [reply index.js:130]: if rollback fails and you try it again, it will most likely fail again. so retrying a rollback is not useful. rather, we should make sure that failing a rollback due to an error happened in the handler returns an error to the client which reflects the error that was thrown in the handler, whatever it was

File: index.js
Comment (simoneb) [reply index.js:138]: if the route has no option to require the use of the transaction, all the plugin code should be a no-op basically

Comment (toni-sharpe) [reply index.js:138]: In order to make this work I had to re-instate these lines from the earlier attempt which ensure a client is available on the `req`:

```
  if (!fastify.hasRequestDecorator('pg')) {
    fastify.decorateRequest('pg', null)
    fastify.addHook('onRequest', async (req, reply) => {
      req.pg = fastify.pg
    })
  }
```

Comment (simoneb) [reply index.js:138]: doesn't look like this code is in any way related to what you're trying to do here. if you need a client, you get it from the pool, but only if you need it

Comment (toni-sharpe) [reply index.js:138]: In the original issue description `pg` is added to the `req` but if we only do that conditionally in `useTransaction` then we don't have that available on the `req` anymore. This makes the API inconsistent, ie, if I change `useTransaction` from true to false I then also have to change from `req.pg` to `fastify.pg` in my code. My useTransaction=false tests failed and enabled me to catch this.

File: index.js
Comment (toni-sharpe) [reply index.js:130]: Does Fastify already do that? ie. `onError` intercepts an error that was already heading to the user via `500` mechanisms?

https://www.fastify.io/docs/latest/Hooks/#onerror

The only option would seem to be throwing a 'rollback failed' 

Wondering what happens to unfinished transactions who's clients are released?

File: add-handler.js
Comment (simoneb) [inline add-handler.js:15]: remove handlers variable and early return

File: index.js
Comment (simoneb) [inline index.js:110]: pull this at the module root

File: index.js
Comment (simoneb) [inline index.js:108]: this should be as suggested in the issue `{ pg: { transact: true } }`

File: package.json
Comment (climba03003) [inline package.json:9]: It should not have `--fix` parameter for `standard`, because it can not ensure the commit files is linted.
```suggestion
    "test": "standard && tap -J test/*.test.js && npm run test:typescript",
```

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: index.js
Comment (simoneb) [inline index.js:128]: this is preHandler, let's define it before onError

File: index.js
Comment (simoneb) [inline index.js:126]: use decorateRequest

File: index.js
Comment (simoneb) [inline index.js:127]: awaiting at the end of an async function is an anti pattern https://www.nearform.com/blog/javascript-promises-the-definitive-guide/

File: index.js
Comment (simoneb) [inline index.js:122]: please check in a unit test where you mock req.pg.query what happens if sending the ROLLBACK query fails, because according to the fastify docs, the onError hook does not support sending an error to the done callback

```suggestion
      const onError = (req, reply, error, done) => {
        req[transactionFailedSymbol] = true

        req.pg.query('ROLLBACK', done)
      }
```

File: index.js
Comment (simoneb) [inline index.js:132]: nit: we may not need the reply and payload args if we're not using them

File: index.js
Comment (simoneb) [inline index.js:139]: no need to rollback if the commit failed. this keeps the same semantics as the transaction util defined above

File: index.js
Comment (simoneb) [inline index.js:156]: let remove this for the time being

Comment (simoneb) [reply index.js:156]: if I were to write this code, I would keep it like: (not 100% sure though)

```suggestion
  if (!fastify.hasRequestDecorator('pg')) {
    fastify.decorateRequest('pg', fastify.pg)
  }
```

**Review by simoneb — CHANGES_REQUESTED** [review]

File: package.json
Comment (toni-sharpe) [reply package.json:9]: I added this because on test I was getting lots of small lint errors so it seemed convenient to have `--fix` fix those for me when running test

How do I run `standard --fix` from the command line?

File: index.js
Comment (simoneb) [inline index.js:124]: the error handling here should be done differently from how it's done in the plugin root. calling next with an Error was fine there, it is not here. Simply throw the error instead. Same at line 130.

File: index.js
Comment (simoneb) [inline index.js:136]: if the plugin is registered with a name, you have to use it everywhere. this will not work. same in all other places where you're accessing req.pg.  all those accesses, in case a name is provided, should be req.pg[name]

File: index.js
Comment (simoneb) [inline index.js:121]: you can't apply the same logic that was applied before, it doesn't make much sense here

File: index.js
Comment (simoneb) [inline index.js:123]: this also doesn't make much sense, you must check against req.pg, not fastify.pg

**Review by simoneb — CHANGES_REQUESTED** [review]

File: index.js
Comment (simoneb) [inline index.js:125]: this works but it's not great to assign it an empty object the populate a property on line 132. just do this on line 132:

```js
req.pg = { [name]: client }
```

Comment (simoneb) [reply index.js:125]: actually no, ignore the above. you can't unconditionally set req.pg to an empty object, because another registration of the plugin may have already set something on it. you must conditionally initialize it if it's not yet initialized

File: index.js
Comment (simoneb) [inline index.js:124]: are you sure about the namespacing logic? you must handle all combinations of:

- non namespaced / namespaced registration
- non namespaced / namespaced route transaction

to make an example of something that doesn't work correctly here, if you register with name `foo` and enable the transaction with name `bar`, this code should do nothing.

File: index.js
Comment (toni-sharpe) [reply index.js:126]: We resolved this in Slack. It makes sense to be defensive and avoid a namespace that matches a property on the `client` object from pg.

File: index.js
Comment (toni-sharpe) [reply index.js:124]: Good spot.

An error might be good in this case too, but I should also make a test that covers it.

Currently working through the test for duplicate namespaces that might have exposed something else, so I will work on this next.

Comment (toni-sharpe) [reply index.js:124]: > if you register with name foo and enable the transaction with name bar, this code should do nothing.

I think this can be handled just by checking the `req` for the correct namespace and if it doesn't exist throwing an error.

Either way, I think we should do this, so I will add an error and test for this

File: index.js
Comment (toni-sharpe) [inline index.js:137]: Must use `Object.assign` here

Comment (toni-sharpe) [reply index.js:148]: In fact, this is caught by the previous line

File: index.js
Comment (simoneb) [inline index.js:161]: you don't need this here, you can simply use client

File: test/req-initialization.test.js
Comment (simoneb) [inline test/req-initialization.test.js:218]: I believe this could be simply `fastify.inject('/')`

File: index.js
Comment (simoneb) [inline index.js:60]: invert the condition here to return early, plus use a more obvious condition: `typeof transact === 'string'`

**Review by simoneb — CHANGES_REQUESTED** [review]

**Review by simoneb — APPROVED** [review]

**Comment by toni-sharpe** [comment]
Hey everybody,

We're ready to take this out of draft and get reviews from other users.

File: add-handler.js
Comment (mcollina) [inline add-handler.js:1]: use strict is missing.

Maybe it's better to move this file inside `lib/`

File: package.json
Comment (mcollina) [inline package.json:8]: why this? Can you remove?

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work! This is missing docs

File: add-handler.js
Comment (jsumners) [inline add-handler.js:1]: Nit:

```suggestion
module.exports = function addHandler (existingHandler, newHandler) {
```

**Review by jsumners — COMMENTED** [review]
Looking good. Missing documentation.

File: package.json
Comment (toni-sharpe) [reply package.json:8]: I found this useful because when I was developing I frequently wanted to see if I'd broken tests. When running that I didn't want something like indentation linting to stop the tests running.

Happy to remove, but that was my reasoning.

Comment (mcollina) [reply package.json:8]: ok

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
cc @voxpelli

File: README.md
Comment (jsumners) [inline README.md:256]:
```suggestion
Important: rolling back a transaction relies on the handler failing and being caught by an `onError` hook. This means that the transaction wrapped route handler must not catch any errors internally.
```

**Review by jsumners — APPROVED** [review]

**Review by simoneb — APPROVED** [review]

**Review by climba03003 — APPROVED** [review]
Actually, I did not use this `plugin` in any of my project. But I do not want to see this blocked by `pending reviewer`. So, I will approve the change.

Overall, when I scan through the code. It looks good.

**Pull request merged**
